samples/reactjs-todo/client/components/journey/web-authn.js

```diff
@@ -0,0 +1,66 @@
+/*
+ * forgerock-sample-web-react
+ *
+ * webAuthn.js '../ components/icons/finger-print-icon'
+ *
+ * Copyright (c) 2021 ForgeRock. All rights reserved.
+ * This software may be modified and distributed under the terms
+ * of the MIT license. See the LICENSE file for details.
+ */
+import { FRWebAuthn, WebAuthnStepType } from '@forgerock/javascript-sdk';
+import React, { useEffect, useState, useContext } from 'react';
+import { AppContext } from '../../global-state';
+/**
+ * @function WebAuthn - Footer React component
+ * @returns {Object} - React component object
+ */
+export default function WebAuthn({ step, setSubmissionStep }) {
+  const [contextState] = useContext(AppContext);
+  const webAuthnStep = FRWebAuthn.getWebAuthnStepType(step);
+  const [state, setState] = useState({
+    message: '',
+    header: '',
+  });
+  useEffect(() => {
+    async function performWebAuthn() {
+      try {
+        if (webAuthnStep === WebAuthnStepType.Registration) {
+          setState({
+            header: 'Registering your device',
+            message: 'Your device will be used to verify your identity',
+          });
+          await FRWebAuthn.register(step);
+        } else {
+          setState({
+            header: 'Verifying your identity',
+            message: 'Use your device to verify your identity',
+          });
+          await FRWebAuthn.authenticate(step);
+        }
+        setSubmissionStep(step);
+      } catch (e) {
+        setSubmissionStep({
```

Comment:
If we set this as the submission step, should we not have something handling this type of error in the `journey` hook? 

What is the expected behavior when someone cancels a webauthn step? A webauthn node has a `unsupported, `success`, `failure` and `client error`  possible case. In the case of cancel, is this `failure`?

if we fail should we be redirecting to a default journey? Is this what happens when we throw this error? I feel like we're missing the handling of this error and just throwing it to throw rather than handling it. I don't believe we have a standard error like this that we handle by default

Reply:
You are correct, Ryan. We should submit the step as is — e.g. (`setSubmissionStep(step)`) — just like in the success case. The `step` gets modified in the WebAuthn method before it throws. This gives the dev an opportunity to handle the error, or just submit it as is.

The client itself never declares `LoginFailure`, that's the responsibility of the server. So, we shouldn't ever directly use this `LoginFailure` step type.